❇️  Modularize step 2 of preprocessing (AFP)

truenomad opened the pull request:
Closes [#161](https://github.com/nish-kishore/tidypolis/issues/161)

This pull request modularizes Step 2 of AFP preprocessing into standalone functions. It:

1. Refactors Step 2 into reusable, documented functions (all prefixed with `s2_`)

1. Introduces `find_latest_archive()` to handle archive lookup (potentially useful for other steps as well).

Tested and runs successfully on my end. @mcuadera & @nish-kishore please have a look and see what you thin... after all it is my first proper pull request on tidypolis :D !

mcuadera reviewed (changes requested):
Hey Mo, everything ran on my end! Step 2 is now much easier to follow and also benefits from better memory management. Here are my comments. There were three minor things I identified that required changes, but the rest are just notes for myself to refer back to:

**General notes:**
For step 2, the following files are prerequisites (in case we want to turn the entirety of step 2 into a function):
1. Path to the POLIS data folder
2. Path to the POLIS folder
3. Long district spatial file (loaded initially in step 1 for processing sub-activity data)

**Comments:**
1. Potentially, `s2_read_afp_data()` and `find_latest_archive()` can be used for Steps 2-5 as it's a general function.

**Change requests:**
1. For `s2_process_coordinates()`, upon loading `shape_data`, we are calling to a path outside of the current value of `polis_data_folder`, which is `"POLIS/data"` or what is in the global variable `POLIS_DATA_CACHE`. Paths in functions should at least be relative paths to any paths passed as an argument, but I understand why we had to do it this way given the only parameter passed to `preprocess_cdc()` is `polis_data_folder`. My suggestion: 
    - Change the parameter to `polis_folder = Sys.getenv("POLIS_DATA_FOLDER")` and then put `polis_data_folder = file.path(polis_folder, "data")` in the first line within `preprocess_cdc()`. So it would now look like:
    ```
    preprocess_cdc(polis_folder = Sys.getenv("POLIS_DATA_FOLDER")) {
      polis_data_folder = file.path(polis_folder, "data")
     # rest of the code ...
    }
     ```
    We could also assign a path to the `misc` folder explicitly as well following the same logic. We can refactor the rest later. Thoughts @nish-kishore @Heaghneyn?
2. Run `devtools::document()` to make sure all the private functions are internally documented. 
3. Add an extra step to make sure `afp_empty_coords.csv` are exported as well. Perhaps just before `Step 2i`. In our old code, this is line [2944](https://github.com/nish-kishore/tidypolis/blob/aa9e8d1accbf203fa137067947a2704a1bb48afd/R/utils.R#L2944C1-L2951C24).
4. Make the combining of historical data optional. At the very end of `s2_export_afp_outputs()`, we check for our static afp files in an if statement. We also need to make sure that if the static files are absent, a non-combined dataset (`afp_new`) and a "light" version gets exported to Core_Ready_Files (line 6966).

mcuadera commented:
One more request I have is making the combining of historical files optional, see above comment on specific on how to implement

nish-kishore commented:
I retcon-ed Step 0 and 1 along with some memory optimization and silencing of progress bars, will continue with Step 2 ASAP

truenomad commented:
Great timing with the push @nish-kishore 

Thanks again for the super constructive feedback, @mcuadera. I’ve now implemented the following changes:

- Reworked `preprocess_cdc()` to take polis_folder and derive `polis_data_folder` internally
- Exported `afp_empty_coords.csv` within `s2_process_coordinates()`
- Ran `devtools::document()` to update internal function documentation
- Wrapped the full Step 2 workflow into a single function: `s2_fully_process_afp_data()`
- Replaced `do.call(rbind, ...)` with `data.table::rbindlist()` to speed up the binding of `long.global.dist.01`

Managed to run this locally without any problem. 

One observation I’ve made is that the two major bottlenecks in Step 2 seem to be the spatial join of AFP cases to global districts, as well as the export process, particularly the part that creates the combined AFP dataset.

Would love to get your thoughts on this, @nish-kishore.

mcuadera commented:
@truenomad and @nish-kishore,  the step within f.pre.stsample.01() that takes the longest is during the st_make_valid() step. Nick and I talked a little bit about it last Friday. We probably don't need to make every district shape files valid and the number of invalid shapes that we have is small relative to the global district shape file . What I'm unsure of is whether the checking part takes the longest vs the making valid part. If it's the checking part, not sure if there's a good way to speed things up.

mcuadera commented:
@truenomad I agree with you that it makes sense to put the spatial checking in the spatial preprocessing function. If so, probably unnecessary to do the logging and correcting within f.pre.st.sample.01(). However, for now, since the spatial preprocessing doesn't have it, let leave it there for now. @Heaghneyn, what do you think?

In a perfect world I would be happy to switch to parquet instead of rds. Parquet has the advantage of not loading data directly into memory, but overall seems to be slightly larger than an RDS file based on my analysis, but that's not a big deal IMO as storage is cheap. RDS has to be loaded directly to memory which is annoying.

However, the main limitation I encountered is that `edav_io()` can't read parquet files directly from EDAV to R, so `get_all_polio_data()` will fail if it's not pulling data locally. Essentially, we need to first download the file into a temp file and then connect the parquet folder to R. Also, our preprocessing system, including archives is built around reading RDS and comparing RDS files so I am hesitant to implement parquet at this time. Needs further discussion and more thorough planning as switching to parquet is going to be a big change in preprocessing.

nish-kishore commented:
@mcuadera, also let's not forget that the size of the object matters quite a bit in terms of utility of the system (for CDC) when we're in the field

nish-kishore commented:
@truenomad @mcuadera @Heaghneyn, I think we're at a point where a quick 30 min call to reach some decisions here would be worthwhile. Want to make sure the score on this PR doesn't creep.

truenomad commented:
> @truenomad, @mcuadera and I went ahead and reverted the last two changes. I think they are both necessary but out of the testing scope of this PR. Some of the key issues we need to still figure out about the optimization and the use of parquet:
> 
> Parquet:
> 
>     * The compression is really interesting and something we need to test out further. Not just in how it stores locally but how it interacts with systems like EDAV. A simple solution would be to add this as a flag in `init_tidypolis` to determine if it stores as RDS or Parquet or even just have it flag depending on if EDAV = T
> 
>     * Another part of the parquet mystery is that I don't really understand that significant of a drop in size without having a key-indexed hierarchy. Generally, objects that get compressed that much should be taking a long time to read in and out (if we're comparing 3.3 GB to 130MB). If it doesn't something is really not making sense.
> 
> 
> AFP optimization:
> 
>     * Also really important. The challenge is incorporating data.table just here. Agree that it is much faster for some tasks but we should verify how much it messes up our dependency graph. Generally I don't like to use a thing in only one place unless absolutely necessary because R (unfortunately) will load the entire library instead of just that function.
> 
>     * the `parallel` function in `data.table` is going to work on iOS and linux but not on Windows machines. We'll need to find another way to deal with that.
> 
> 
> We have the edits on record and they're not massive so we can go back and incorporate them again after we figure out the best way forward.

Thanks for this @nish-kishore, and understood in terms of scope. I’ll put together a separate PR for it!

truenomad commented on R/utils.R:
```diff
@@ -2305,4173 +2305,4735 @@ preprocess_cdc <- function(polis_data_folder = Sys.getenv("POLIS_DATA_CACHE")) {
 
-    cli::cli_alert_info("No previous archive identified, will not perform any comparisons")
+  # read and process AFP data
+  afp_raw_new <- s2_read_afp_data(y[grepl("Human", y)])
```
Yes! Sounds good with me @nish-kishore. Certainly helps with the debugging given that we have so many functions for the preprocessing.
